crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,343 @@ impl MeshletMesh {
+    }
+}
+
+fn compute_meshlets(indices: &[u32], vertices: &VertexDataAdapter) -> Meshlets {
+    let mut meshlets = build_meshlets(indices, vertices, 64, 64, 0.0);
+
+    for i in 0..meshlets.len() {
+        let meshlet = meshlets.meshlets[i];
+        #[allow(unsafe_code)]
+        #[allow(clippy::undocumented_unsafe_blocks)]
+        unsafe {
+            meshopt_optimizeMeshlet(
+                &mut meshlets.vertices[meshlet.vertex_offset as usize],
+                &mut meshlets.triangles[meshlet.triangle_offset as usize],
+                meshlet.triangle_count as usize,
+                meshlet.vertex_count as usize,
+            );
+        }
+    }
+
+    meshlets
+}
+
+fn collect_triangle_edges_per_meshlet(
+    simplification_queue: Range<usize>,
+    meshlets: &Meshlets,
+) -> HashMap<usize, HashSet<(u32, u32)>> {
+    let mut triangle_edges_per_meshlet = HashMap::new();
+    for meshlet_id in simplification_queue {
+        let meshlet = meshlets.get(meshlet_id);
+        let meshlet_triangle_edges = triangle_edges_per_meshlet
+            .entry(meshlet_id)
+            .or_insert(HashSet::new());
+        for i in meshlet.triangles.chunks(3) {
+            let v0 = meshlet.vertices[i[0] as usize];
+            let v1 = meshlet.vertices[i[1] as usize];
+            let v2 = meshlet.vertices[i[2] as usize];
+            meshlet_triangle_edges.insert((v0.min(v1), v0.max(v1)));
+            meshlet_triangle_edges.insert((v0.min(v2), v0.max(v2)));
+            meshlet_triangle_edges.insert((v1.min(v2), v1.max(v2)));
+        }
+    }
+    triangle_edges_per_meshlet
+}
+
+fn find_connected_meshlets(
+    simplification_queue: Range<usize>,
+    triangle_edges_per_meshlet: &HashMap<usize, HashSet<(u32, u32)>>,
+) -> HashMap<usize, Vec<(usize, usize)>> {
+    let mut connected_meshlets_per_meshlet = HashMap::new();
+    for meshlet_id in simplification_queue.clone() {
+        connected_meshlets_per_meshlet.insert(meshlet_id, Vec::new());
+    }
+
+    for (meshlet_id1, meshlet_id2) in simplification_queue.tuple_combinations() {
+        let shared_edge_count = triangle_edges_per_meshlet[&meshlet_id1]
+            .intersection(&triangle_edges_per_meshlet[&meshlet_id2])
+            .count();
+        if shared_edge_count != 0 {
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id1)
+                .unwrap()
+                .push((meshlet_id2, shared_edge_count));
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id2)
+                .unwrap()
+                .push((meshlet_id1, shared_edge_count));
+        }
+    }
+    connected_meshlets_per_meshlet
+}
+
+fn group_meshlets(
+    simplification_queue: Range<usize>,
+    connected_meshlets_per_meshlet: &HashMap<usize, Vec<(usize, usize)>>,
+) -> HashMap<i32, Vec<usize>> {
+    let mut xadj = Vec::with_capacity(simplification_queue.len() + 1);
+    let mut adjncy = Vec::new();
+    let mut adjwgt = Vec::new();
```

Comment:
I think I'd prefer some more descriptive names here. Also a comment explaining what's going on, because the algorithm isn't obvious to me :)

Reply:
METIS takes a very specific input format. The names are copied from what METIS expects. I linked the METIS manual, read that to understand how it works. Comments wouldn't really help.

Reply:
Could a link to the relevant section of METIS be added here?